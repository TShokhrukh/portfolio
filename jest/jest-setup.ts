import '@testing-library/jest-dom'
import { TextEncoder, TextDecoder } from 'util'
import { IntersectionObserverMock } from './mock/IntersectionObserverMock '

// @ts-ignore
global.TextEncoder = TextEncoder
// @ts-ignore
global.TextDecoder = TextDecoder
// @ts-ignore
global.IntersectionObserver = IntersectionObserverMock
