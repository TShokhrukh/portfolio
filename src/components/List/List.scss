.List {
  --list-gap: #{$spacer};

  display: grid;
  gap: var(--list-gap);
}
