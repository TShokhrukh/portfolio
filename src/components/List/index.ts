export * from './build'
export * from './__types__'
export { List as ListBase } from './List'
