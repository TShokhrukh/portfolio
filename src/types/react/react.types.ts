import type { ElementType } from 'react'

export interface IComponentProp<P = any> {
  component?: ElementType<P>,
}
