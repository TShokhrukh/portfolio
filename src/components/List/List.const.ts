import { createContext } from 'react'
import { cn } from '@bem-react/classname'
import { IListContext } from './__types__/List.types'

export const cnList = cn('List')

export const Context = createContext<IListContext>(null!)
