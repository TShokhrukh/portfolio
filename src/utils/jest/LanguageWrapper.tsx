import React, { ReactNode } from 'react'
import { TranslationsProvider } from '@eo-locale/react'

import type { Locale } from '@eo-locale/core'

interface IWrapperProps {
  children: ReactNode,
  language?: string,
  locales?: Locale[]
}

const defLocales: Locale[] = [
  {
    language: 'en',
    messages: {}
  }
]

export const LanguageWrapper: React.FC<IWrapperProps> = ({
  children,
  language = 'en',
  locales = defLocales
}) => {
  return (
    <TranslationsProvider language={language} locales={locales} onError={() => null}>
      {children}
    </TranslationsProvider>
  )
}
