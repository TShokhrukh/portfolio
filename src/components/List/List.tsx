import React, { FC } from 'react'
import { Context, cnList } from './List.const'
import { ListNoItems } from './NoItems'
import './List.scss'
import type { IListProps } from './__types__'

export const List: FC<IListProps> = ({
  items,
  className,
  component: Component = 'div',
  itemComponent: ItemComponent = 'div',
  itemPropKey,
  noItemsText = 'list.noItems',
  ...props
}) => {
  return (
    <Context.Provider value={{ noItemsText }}>
      <Component {...props} className={cnList({ empty: !items.length }, [className])}>
        {items.length
          ? items.map(i => <ItemComponent key={i.id} {...(itemPropKey ? { [itemPropKey]: i } : i)} />)
          : <ListNoItems />}
      </Component>
    </Context.Provider>
  )
}
