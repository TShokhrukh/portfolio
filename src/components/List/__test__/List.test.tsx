/* global test, expect, describe, jest */
import React from 'react'
import { render, screen } from '@testing-library/react'
import { LanguageWrapper } from '~/utils/jest/LanguageWrapper'
import { List } from '../List'

describe('List', () => {
  const items = [
    {
      id: '1',
      label: 'item-1'
    },
    {
      id: '2',
      label: 'item-2'
    }
  ]

  test('must return the component template', () => {
    render(
      <LanguageWrapper>
        <List items={items} data-testid='list' itemComponent={({ id }) => <div>item-{id}</div>} />
      </LanguageWrapper>
    )

    expect(screen.getByTestId('list')).toMatchSnapshot()
  })

  test('must display "no items message" if there are no items', () => {
    render(
      <LanguageWrapper>
        <List items={[]} />
      </LanguageWrapper>
    )

    expect(screen.getByText('list.noItems')).toBeInTheDocument()
  })

  test('prop "noItemsText" must change "no items message"', () => {
    render(
      <LanguageWrapper>
        <List items={[]} noItemsText='noItemsText' />
      </LanguageWrapper>
    )

    expect(screen.getByText('noItemsText')).toBeInTheDocument()
  })

  test('must pass the list item to the component', () => {
    const SpyComponent = jest.fn()

    render(
      <LanguageWrapper>
        <List items={items} itemComponent={SpyComponent} />
      </LanguageWrapper>
    )

    expect(SpyComponent).toHaveBeenCalledTimes(2)
    items.forEach((i, n) => {
      expect(SpyComponent).toHaveBeenNthCalledWith(n + 1, i, {})
    })
  })

  test('must pass the list item in the property set in the "itemPropKey"', () => {
    const SpyComponent = jest.fn()

    render(
      <LanguageWrapper>
        <List items={items} itemPropKey='item' itemComponent={SpyComponent} />
      </LanguageWrapper>
    )

    expect(SpyComponent).toHaveBeenCalledTimes(2)
    items.forEach((i, n) => {
      expect(SpyComponent).toHaveBeenNthCalledWith(n + 1, { item: i }, {})
    })
  })
})
