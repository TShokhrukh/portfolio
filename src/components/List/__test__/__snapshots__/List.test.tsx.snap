// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List must return the component template 1`] = `
<div
  class="List"
  data-testid="list"
>
  <div>
    item-
    1
  </div>
  <div>
    item-
    2
  </div>
</div>
`;
